apiserver/facades/controller/charmrevisionupdater/interface_test.go

```diff
@@ -110,94 +105,46 @@ func makeResource(c *gc.C, name string, revision, size int, hexFingerprint strin
 	}
 }
 
-func newFakeCharmhubClient(st charmrevisionupdater.State, metadata map[string]string) (charmrevisionupdater.CharmhubRefreshClient, error) {
-	resources := []transport.ResourceRevision{
-		{
-			Download: transport.ResourceDownload{
-				HashSHA384: "59e1748777448c69de6b800d7a33bbfb9ff1b463e44354c3553bcdb9c666fa90125a3c79f90397bdf5f6a13de828684f",
-				Size:       5,
-			},
-			Name:     "reza",
-			Revision: 7,
-			Type:     "file",
-		},
-		{
-			Download: transport.ResourceDownload{
-				HashSHA384: "03130092073c5ac523ecb21f548b9ad6e1387d1cb05f3cb892fcc26029d01428afbe74025b6c567b6564a3168a47179a",
-				Size:       6,
-			},
-			Name:     "rezb",
-			Revision: 1,
-			Type:     "file",
-		},
-	}
-	charms := map[string]charmhubCharm{
-		"charm-1": {id: "charm-1", name: "mysql", revision: 23},
-		"charm-2": {id: "charm-2", name: "postgresql", revision: 42},
-		"charm-3": {id: "charm-3", name: "resourcey", revision: 1, resources: resources},
-	}
-	return &fakeCharmhubClient{charms: charms, metadata: metadata}, nil
+// charmhubConfigMatcher matches only the charm IDs and revisions of a
+// charmhub.RefreshMany config.
+type charmhubConfigMatcher struct {
+	expected []charmhubConfigExpected
 }
 
-type charmhubCharm struct {
-	id        string
-	name      string
-	revision  int
-	resources []transport.ResourceRevision
-	version   string
-}
-
-type fakeCharmhubClient struct {
-	charms   map[string]charmhubCharm
-	metadata map[string]string
+type charmhubConfigExpected struct {
+	id       string
+	revision int
 }
 
-func (c *fakeCharmhubClient) Refresh(_ context.Context, config charmhub.RefreshConfig) ([]transport.RefreshResponse, error) {
-	// Sanity check that metadata headers are present
-	if c.metadata["model_uuid"] == "" {
-		return nil, errors.Errorf("model metadata not present")
+func (m charmhubConfigMatcher) Matches(x interface{}) bool {
+	config, ok := x.(charmhub.RefreshConfig)
+	if !ok {
+		return false
```

Comment:
I've started to add c *gc.C to the matchers, then I can do
 c.Assert(ok, gc.IsTrue, gc.Commentf("detailed reason")
You can stop the test there, and know exactly what the issue is, rather than the terse go mock response.

Reply:
I think this would be fine for the interface type conversion (as that's simply a bug in the code). Though I'm not sure you can do this in general, because the `Matches()` method has to return false if it doesn't match in case of matching multiple items. I decided to just leave this as is.

Reply:
For future reference: Ian noted that you can `c.Fail()` and still return false.